### PR TITLE
HPCC-14000 Better fix for incorrect balanced splitters

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday gavin.halliday@lexisnexis.com

### CODE REVIEW

**Comment by ghalliday** [comment]
@richardkchapman I think it might be worth talking through this problem to see if anyone has any other bright ideas when we next meet.  It will be easier to discuss in person.
@jakesmith @jamienoss please review.

**Comment by jamienoss** [comment]
@ghalliday I guess we forgot to look at this at the offsite...

**Comment by ghalliday** [comment]
@richardkchapman please review

File: ecl/hqlcpp/hqlresource.cpp [hunk 1]
```diff
@@ -5273,10 +5281,119 @@ IHqlExpression * EclResourcer::walkPotentialSplitterLinks(CSplitterInfo & connec
...
+#endif
+
+    sourceLink->mergeSinkLink(*sinkLink);
+    return true;
+}
+
+void EclResourcer::removeDuplicateIndependentLinks(CSplitterInfo & connections, ResourcerInfo & info)
+{
+    IHqlExpression * expr = info.original;
+    bool again = true;
```
Comment (richardkchapman) [inline ecl/hqlcpp/hqlresource.cpp]: This feels like an odd way of coding a do-wihle loop...

File: ecl/hqlcpp/hqlresource.cpp [hunk 2]
```diff
@@ -5273,10 +5281,119 @@ IHqlExpression * EclResourcer::walkPotentialSplitterLinks(CSplitterInfo & connec
...
+#ifdef TRACE_BALANCED
+                    if (numRemoved)
+                        printf("//removed %u duplicate links from %u to %u\n", numRemoved, info.balanceId, sinkInfo.balanceId);
+#endif
+
+                }
+
+                //Removing duplicate links has turned the source item into a pass-through
+                //replace references to the sink activity with references to its sink
+                //to possible allow more to be removed
```
Comment (richardkchapman) [inline ecl/hqlcpp/hqlresource.cpp]: Grammar - possible -> possibly ?

File: ecl/hqlcpp/hqlresource.cpp [hunk 3]
```diff
@@ -5273,10 +5281,119 @@ IHqlExpression * EclResourcer::walkPotentialSplitterLinks(CSplitterInfo & connec
...
+
+#ifdef TRACE_BALANCED
+                    if (numRemoved)
+                        printf("//removed %u duplicate links from %u to %u\n", numRemoved, info.balanceId, sinkInfo.balanceId);
+#endif
+
+                }
+
+                //Removing duplicate links has turned the source item into a pass-through
+                //replace references to the sink activity with references to its sink
```
Comment (richardkchapman) [inline ecl/hqlcpp/hqlresource.cpp]: Capital R on replace would make it slightly easier to read

File: ecl/hqlcpp/hqlresource.cpp [hunk 4]
```diff
@@ -5273,10 +5281,119 @@ IHqlExpression * EclResourcer::walkPotentialSplitterLinks(CSplitterInfo & connec
...
+    bool again = true;
+    while (again)
+    {
+        again = false;
+        for (unsigned i=0; i < info.balancedLinks.ordinality(); i++)
+        {
+            CSplitterLink & cur = info.balancedLinks.item(i);
+            if (cur.hasSource(expr))
+            {
+                IHqlExpression * sink = cur.queryOther(expr);
```
Comment (richardkchapman) [inline ecl/hqlcpp/hqlresource.cpp]: No chance of NULL returns here?

**Comment by richardkchapman** [comment]
@ghalliday Some minor questions/comments

File: ecl/hqlcpp/hqlresource.cpp (hunk 4 above)
Comment (ghalliday) [inline ecl/hqlcpp/hqlresource.cpp]: No, but I'll add a check.

**Comment by ghalliday** [comment]
See #8001 for new version

**Pull request closed without merging**